gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java

```diff
@@ -294,39 +373,357 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
     }
   }
 
-  private void processJob(State jobProps) {
-    this.futures.add(this.executorService.submit(getMRCompactorJobRunner(jobProps)));
+  private void processCompactionJobs() throws IOException {
+    if (this.shouldVerifDataCompl) {
+      verifyDataCompleteness();
+    } else {
+      setAllDatasetStatesToVerified();
+    }
+
+    this.submitCompactionJobsAndWaitForCompletion();
+  }
+
+  private boolean shouldVerifyDataCompleteness() {
+    return this.state.getPropAsBoolean(ConfigurationKeys.COMPACTION_SHOULD_VERIFY_DATA_COMPLETENESS,
+        ConfigurationKeys.DEFAULT_COMPACTION_SHOULD_VERIFY_DATA_COMPLETENESS);
+  }
+
+  private void verifyDataCompleteness() {
+    List<Pattern> blacklist = getDataComplVerifBlacklist();
+    List<Pattern> whitelist = getDataComplVerifWhitelist();
+    int numDatasetsVerifiedTogether = getNumDatasetsVerifiedTogether();
+    List<Dataset> datasetsToBeVerified = Lists.newArrayList();
+    for (Dataset dataset : this.datasets) {
+      if (DatasetFilterUtils.survived(getTopic(dataset.jobProps()), blacklist, whitelist)) {
+        datasetsToBeVerified.add(dataset);
+        if (datasetsToBeVerified.size() >= numDatasetsVerifiedTogether) {
+          ListenableFuture<Results> future = this.verifier.verify(datasetsToBeVerified);
+          addCallback(datasetsToBeVerified, future);
+          datasetsToBeVerified = Lists.newArrayList();
+        }
+      } else {
+        dataset.setState(VERIFIED);
+      }
+    }
+  }
+
+  private void addCallback(final List<Dataset> datasetsToBeVerified, ListenableFuture<Results> future) {
+
+    Futures.addCallback(future, new FutureCallback<Results>() {
+
+      /**
+       * On success, resubmit verification for the {@link Dataset}s that should be resubmitted
+       * (i.e., verification didn't pass and it didn't timeout).
+       */
+      @Override
+      public void onSuccess(Results results) {
+        List<Dataset> datasetsToBeVerifiedAgain = Lists.newArrayList();
+        for (Results.Result result : results) {
+          Optional<MRCompactorJobRunner> jobRunner =
+              Optional.fromNullable(MRCompactor.this.jobRunnables.get(result.dataset()));
+
+          if (result.status() == DataCompletenessVerifier.Results.Result.Status.PASSED) {
```

Comment:
I think it is better to have a switch statement here. If someone later adds a new state, there will be a warning that there is a case that has not been processed.